src/proxy/http/profiles.rs

```diff
@@ -295,49 +311,55 @@ pub mod router {
     }
 
     #[derive(Debug)]
-    pub struct Layer<G, M, R, B> {
+    pub struct Layer<G, M, R, RBody, MBody> {
         get_routes: G,
         route_layer: svc::Builder<R>,
         suffixes: Vec<dns::Suffix>,
         /// This is saved into a field so that the same `Arc`s are used and
         /// cloned, instead of calling `Route::default()` every time.
         default_route: Route,
-        _p: ::std::marker::PhantomData<fn() -> (M, B)>,
+        _p: ::std::marker::PhantomData<fn() -> (M, RBody, MBody)>,
     }
 
     #[derive(Debug)]
-    pub struct MakeSvc<G, M, R, B> {
+    pub struct MakeSvc<G, M, R, RBody, MBody> {
         inner: M,
         get_routes: G,
         route_layer: svc::Builder<R>,
         suffixes: Vec<dns::Suffix>,
         default_route: Route,
-        _p: ::std::marker::PhantomData<fn(B)>,
+        _p: ::std::marker::PhantomData<fn(RBody, MBody)>,
     }
 
-    pub struct Service<G, T, R, B>
+    pub struct Service<G, T, R, S, RMk, M, RBody>
     where
         T: WithRoute + Clone,
         T::Output: Eq + Hash,
-        R: rt::Make<T::Output>,
-        R::Value: svc::Service<http::Request<B>> + Clone,
+        R: svc::Layer<S, Service = RMk>,
+        RMk: rt::Make<T::Output>,
+        RMk::Value: svc::Service<http::Request<RBody>> + Clone,
     {
         target: T,
-        stack: R,
+        inner: M,
+        route_layer: svc::Builder<R>,
         route_stream: Option<G>,
-        router: Router<B, T, R>,
+        router: rt::Router<http::Request<RBody>, RouteRecognize<T>, RMk>,
         default_route: Route,
+        _p: ::std::marker::PhantomData<S>,
     }
 
-    type Router<B, T, M> = rt::Router<http::Request<B>, Recognize<T>, M>;
-
-    pub struct Recognize<T> {
+    pub struct RouteRecognize<T> {
         target: T,
-        routes: Routes,
+        routes: Vec<(RequestMatch, Route)>,
         default_route: Route,
     }
 
-    impl<B, T> rt::Recognize<http::Request<B>> for Recognize<T>
+    pub struct ConcreteDstRecognize<T> {
```

Comment:
Nit/TIOLI: WDYT of making these `RecognizeRoute`/`RecognizeConcreteDst`? Seems...verb-ier.

Reply:
neither way seems particularly better or worse than the other to me.  I can change it if you feel strongly.